### DISCUSSION TITLE
getting TypeError: source must an AudioSource not str

### DISCUSSION THREAD

**Discussion by bitmvk** [Rapptz/discord.py#6930] [Q&A]
hi github, I was writing code for music bot it ran ok until I decided to add a queue to it. now it plays one time then stops saying TypeError: source must an AudioSource not str

code:
`import discord
import youtube_dl
from discord.ext import commands
import urllib
import re
import pafy
from discord import FFmpegPCMAudio, PCMVolumeTransformer

client = commands.Bot(command_prefix=["!", "/", ">", "."])

FFMPEG_OPTIONS = {
'before_options': '-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5',
'options': '-vn'
}

song_queue = []

@client.event
async def on_ready():
    print("Bot is ready")

@client.command()
async def join(ctx):
    if not ctx.message.author.voice:
        await ctx.send(":x:**You have to be in a voice channel to use this command**")
        return
    else:
        channel = ctx.message.author.voice.channel
    await channel.connect()
    await ctx.send("**Connected to the channel. it is time to rock**:metal:")

@client.command()
async def leave(ctx):
    voice_client = ctx.message.guild.voice_client
    try:
        await voice_client.disconnect()
        await ctx.send("Party over disconnecting from voice channel")
    except:
        await ctx.send(":x:**I am not connected to a voice channel.** Type `!join` to get me into one.")

@client.command()
async def play(ctx, *args):
    if ctx.message.author.voice == None:
        await ctx.send(":x:**You have to be in a voice channel to use this command**")
        return

    channel = ctx.message.author.voice.channel

    voice = discord.utils.get(ctx.guild.voice_channels, name=channel.name)

    voice_client = discord.utils.get(client.voice_clients, guild=ctx.guild)

    if voice_client == None:
        voice_client = await voice.connect()
    else:
        await voice_client.move_to(channel)

    search = '+'.join(args)

    html = urllib.request.urlopen("https://www.youtube.com/results?search_query=" + search)
    video_ids = re.findall(r"watch\?v=(\S{11})", html.read().decode())

    await ctx.send("https://www.youtube.com/watch?v=" + video_ids[0])

    song = pafy.new(video_ids[0])

    audio = song.getbestaudio()

    source = FFmpegPCMAudio(audio.url, **FFMPEG_OPTIONS)

    song_queue.append(source)

    voice_client.play(source, after= lambda e: next(ctx))

def next(ctx):
    voice_client = discord.utils.get(client.voice_clients, guild=ctx.guild)
    song_queue.pop(0)
    source = str(song_queue[0])
    voice_client.play(source, after= lambda e: next(ctx))
    voice_client.is_playing()

    if channel and not voice.is_playing():
        voice.play(song_queue[0], after=lambda e: repeat(ctx.guild, voice, audio))
        voice.is_playing()

@client.command()
async def queue(ctx, *args):
    search = '+'.join(args)

    html = urllib.request.urlopen("https://www.youtube.com/results?search_query=" + search)
    video_ids = re.findall(r"watch\?v=(\S{11})", html.read().decode())

    await ctx.send("https://www.youtube.com/watch?v=" + video_ids[0])

    song = pafy.new(video_ids[0])

    audio = song.getbestaudio()

    source = FFmpegPCMAudio(audio.url, **FFMPEG_OPTIONS)

    song_queue.append(source)

client.run(TOKEN)
`

runs till play() then tries to use next then stops

what is the problem with this  code

**Comment by jelni** [comment]
Please provide a full traceback including the line number where the error occurred. Also, please use code blocks, so the code is easier to read:
\`\`\`py
\<paste your code\>
\`\`\`
You should consider using aiohttp instead of urllib for loading the website asynchronously without blocking.

**Reply by bitmvk** [reply]
will change the code in near future to use aiohttp but now I only know how to use urllib

**Comment by bitmvk** [comment]
```import discord
import youtube_dl
from discord.ext import commands
import urllib
import re
import pafy
from discord import FFmpegPCMAudio, PCMVolumeTransformer

client = commands.Bot(command_prefix=["!", "/", ">", "."])

FFMPEG_OPTIONS = {
'before_options': '-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5',
'options': '-vn'
}

queue = []

@client.event
async def on_ready():
    print("Bot is ready")

@client.command()
async def join(ctx):
    if not ctx.message.author.voice:
        await ctx.send("❌You have to be in a voice channel to use this command")
        return
    else:
        channel = ctx.message.author.voice.channel
        await channel.connect()
        await ctx.send("Connected to the channel. it is time to rock🤘")

@client.command()
async def leave(ctx):
    voice_client = ctx.message.guild.voice_client
    try:
        await voice_client.disconnect()
        await ctx.send("Party over disconnecting from voice channel")
    except:
        await ctx.send("❌I am not connected to a voice channel. Type !join to get me into one.")

@client.command()
async def play(ctx, *args):
    if ctx.message.author.voice == None:
        await ctx.send("❌You have to be in a voice channel to use this command")
    return

    channel = ctx.message.author.voice.channel

    voice = discord.utils.get(ctx.guild.voice_channels, name=channel.name)

    voice_client = discord.utils.get(client.voice_clients, guild=ctx.guild)

    if voice_client == None:
        voice_client = await voice.connect()
    else:
        await voice_client.move_to(channel)

    search = '+'.join(args)

    html = urllib.request.urlopen("https://www.youtube.com/results?search_query=" + search)
    video_ids = re.findall(r"watch\?v=(\S{11})", html.read().decode())

    await ctx.send("https://www.youtube.com/watch?v=" + video_ids[0])

    song = pafy.new(video_ids[0])

    audio = song.getbestaudio()

    source = FFmpegPCMAudio(audio.url, **FFMPEG_OPTIONS)

    queue.append(video_ids[0])

    voice_client.play(source, after= lambda e: next(ctx))

def next(ctx):
    voice_client = discord.utils.get(client.voice_clients, guild=ctx.guild)
    queue.pop(0)
    source = str(queue[0])
    voice_client.play(source, after= lambda e: next(ctx))
    voice_client.is_playing()

    if channel and not voice.is_playing():
        voice.play(queue[0], after=lambda e: repeat(ctx.guild, voice, audio))
        voice.is_playing()
@client.command()
async def queue(ctx, *args):
    search = '+'.join(args)

    html = urllib.request.urlopen("https://www.youtube.com/results?search_query=" + search)
    video_ids = re.findall(r"watch\?v=(\S{11})", html.read().decode())

    await ctx.send("https://www.youtube.com/watch?v=" + video_ids[0])

    queue.append(video_ids[0])
client.run(TOKEN)
```

error traceback:

```Ignoring exception in command queue:
Traceback (most recent call last):
  File "C:\Users\nandu\AppData\Roaming\Python\Python39\site-packages\discord\ext\commands\core.py", line 85, in wrapped
    ret = await coro(*args, **kwargs)
  File "C:\Users\nandu\Desktop\pyProjects\test2bot.py", line 100, in queue
    queue.append(source)
AttributeError: 'Command' object has no attribute 'append'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\nandu\AppData\Roaming\Python\Python39\site-packages\discord\ext\commands\bot.py", line 939, in invoke
    await ctx.command.invoke(ctx)
  File "C:\Users\nandu\AppData\Roaming\Python\Python39\site-packages\discord\ext\commands\core.py", line 863, in invoke
    await injected(*ctx.args, **ctx.kwargs)
  File "C:\Users\nandu\AppData\Roaming\Python\Python39\site-packages\discord\ext\commands\core.py", line 94, in wrapped
    raise CommandInvokeError(exc) from exc
discord.ext.commands.errors.CommandInvokeError: Command raised an exception: AttributeError: 'Command' object has no attribute 'append'```

**Reply by jelni** [reply]
You've changed the code, so now it tries to `.append()` the video ID to the command `queue` (you have to name the variable something different). Then it will try to play the video ID (a string), which doesn't make sense (you have to download the audio first)

**Reply by bitmvk** [reply]
no you can use  `FFmpegPCMAudio(audio.url, **FFMPEG_OPTIONS) `command to make it play without any downloading required
and I changed the code because I was trying to fix the issue  and dint have the old code with me
